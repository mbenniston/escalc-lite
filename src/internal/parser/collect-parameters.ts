import { builtIns } from '../evaluator/tree-walker'
import type { LogicalExpression } from './logical-expression'

const builtinFunctionNames = new Set(Object.keys(builtIns))

export function collectParameters(expression: LogicalExpression): {
  parameters: Set<string>
  functions: Set<string>
} {
  const parameters = new Set<string>()
  const functions = new Set<string>()
  collect(expression, { parameters, functions })
  return { parameters, functions }
}

function collect(
  expression: LogicalExpression,
  outputs: { parameters: Set<string>; functions: Set<string> },
): void {
  switch (expression.type) {
    case 'value':
      {
        if (expression.value.type === 'parameter') {
          outputs.parameters.add(expression.value.name)
        }
      }
      break
    case 'ternary':
      collect(expression.left, outputs)
      collect(expression.middle, outputs)
      collect(expression.right, outputs)
      break
    case 'binary':
      collect(expression.left, outputs)
      collect(expression.right, outputs)
      break
    case 'unary':
      collect(expression.expression, outputs)
      break
    case 'function':
      expression.arguments.forEach((arg) => {
        collect(arg, outputs)
      })

      if (!builtinFunctionNames.has(expression.name))
        outputs.functions.add(expression.name)
      break
  }
}
