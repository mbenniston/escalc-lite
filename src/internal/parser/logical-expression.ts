export type LogicalExpression =
  | TernaryExpression
  | BinaryExpression
  | UnaryExpression
  | ValueExpression
  | FunctionExpression

export type TernaryExpression = {
  type: 'ternary'
  left: LogicalExpression
  right: LogicalExpression
  middle: LogicalExpression
}

export type BinaryExpression = {
  type: 'binary'
  operator:
    | 'subtraction'
    | 'addition'
    | 'multiplication'
    | 'division'
    | 'modulus'
    | 'exponentiation'
    | 'more-than'
    | 'less-than'
    | 'more-than-equal'
    | 'less-than-equal'
    | 'not-equals'
    | 'equals'
    | 'and'
    | 'or'
    | 'bit-and'
    | 'bit-or'
    | 'bit-xor'
    | 'bit-left-shift'
    | 'bit-right-shift'
    | 'in'
    | 'not-in'
  left: LogicalExpression
  right: LogicalExpression
}

export type UnaryExpression = {
  type: 'unary'
  operator: 'not' | 'bit-complement' | 'negate'
  expression: LogicalExpression
}

export type ValueExpression = {
  type: 'value'
  value:
    | { type: 'constant'; value: unknown }
    | { type: 'parameter'; name: string }
    | { type: 'list'; items: LogicalExpression[] }
}

export type FunctionExpression = {
  type: 'function'
  name: string
  arguments: LogicalExpression[]
}
