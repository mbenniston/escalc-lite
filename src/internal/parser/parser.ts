import { BufferedStream } from '../stream/buffered-stream'
import { CharacterStream } from '../stream/character-stream'
import { Tokenizer } from '../tokenizer/tokenizer'
import { SemanticError } from './errors/semantic-error'
import { UnexpectedTokenError } from './errors/unexpected-token-error'
import { DefaultLiteralFactory, type LiteralFactory } from './literal-factory'
import { match, matchOperators, matchOrThrow, type Scanner } from './utils'
import type {
  BinaryExpression,
  LogicalExpression,
  UnaryExpression,
} from './logical-expression'

const defaultLiteralFactory = new DefaultLiteralFactory()

export function parse(
  expression: string,
  literalFactory: LiteralFactory = defaultLiteralFactory,
): LogicalExpression {
  const scanner = new BufferedStream(
    new Tokenizer(new BufferedStream(new CharacterStream(expression))),
  )

  return logicalExpression(scanner, literalFactory)
}

function logicalExpression(
  scanner: Scanner,
  literalFactory: LiteralFactory,
): LogicalExpression {
  return ternary(scanner, literalFactory)
}

function ternary(
  scanner: Scanner,
  literalFactory: LiteralFactory,
): LogicalExpression {
  let left = or(scanner, literalFactory)

  while (true) {
    const matchedOperator = matchOperators(scanner, ['?'])

    if (matchedOperator !== null) {
      const middle = or(scanner, literalFactory)

      matchOrThrow(scanner, 'colon')

      const right = or(scanner, literalFactory)

      left = { type: 'ternary', left, middle, right }
    } else {
      return left
    }
  }
}

function or(
  scanner: Scanner,
  literalFactory: LiteralFactory,
): LogicalExpression {
  let left = and(scanner, literalFactory)

  const operators = ['||']
  const operatorMap: Record<string, BinaryExpression['operator']> = {
    '||': 'or',
  } as const

  while (true) {
    const matchedOperator = matchOperators(scanner, operators)

    if (matchedOperator !== null && matchedOperator in operatorMap) {
      const operator = operatorMap[matchedOperator]
      const right = and(scanner, literalFactory)
      left = { type: 'binary', operator, left, right }
    } else {
      return left
    }
  }
}
function and(
  scanner: Scanner,
  literalFactory: LiteralFactory,
): LogicalExpression {
  let left = comparison(scanner, literalFactory)

  const operators = ['&&']
  const operatorMap: Record<string, BinaryExpression['operator']> = {
    '&&': 'and',
  } as const

  while (true) {
    const matchedOperator = matchOperators(scanner, operators)

    if (matchedOperator !== null && matchedOperator in operatorMap) {
      const operator = operatorMap[matchedOperator]
      const right = comparison(scanner, literalFactory)
      left = { type: 'binary', operator, left, right }
    } else {
      return left
    }
  }
}

function comparison(
  scanner: Scanner,
  literalFactory: LiteralFactory,
): LogicalExpression {
  let left = bitOr(scanner, literalFactory)

  const operators = ['>', '<', '<=', '>=', '!=', '==', '=', '<>']
  const operatorMap: Record<string, BinaryExpression['operator']> = {
    '>': 'more-than',
    '<': 'less-than',
    '<=': 'less-than-equal',
    '>=': 'more-than-equal',
    '!=': 'not-equals',
    '<>': 'not-equals',
    '==': 'equals',
    '=': 'equals',
  } as const

  while (true) {
    if (matchOperators(scanner, ['not'])) {
      if (matchOrThrow(scanner, 'operator').operator !== 'in') {
        throw new SemanticError('Expected "in" after "not"')
      }

      const items = bitOr(scanner, literalFactory)
      return { type: 'binary', operator: 'not-in', left, right: items }
    }
    if (matchOperators(scanner, ['in'])) {
      const items = bitOr(scanner, literalFactory)
      return { type: 'binary', operator: 'in', left, right: items }
    }

    const matchedOperator = matchOperators(scanner, operators)

    if (matchedOperator !== null && matchedOperator in operatorMap) {
      const operator = operatorMap[matchedOperator]
      const right = bitOr(scanner, literalFactory)
      left = { type: 'binary', operator, left, right }
    } else {
      return left
    }
  }
}

function bitOr(
  scanner: Scanner,
  literalFactory: LiteralFactory,
): LogicalExpression {
  let left = bitXor(scanner, literalFactory)

  const operators = ['|']
  const operatorMap: Record<string, BinaryExpression['operator']> = {
    '|': 'bit-or',
  } as const

  while (true) {
    const matchedOperator = matchOperators(scanner, operators)

    if (matchedOperator !== null && matchedOperator in operatorMap) {
      const operator = operatorMap[matchedOperator]
      const right = bitXor(scanner, literalFactory)
      left = { type: 'binary', operator, left, right }
    } else {
      return left
    }
  }
}

function bitXor(
  scanner: Scanner,
  literalFactory: LiteralFactory,
): LogicalExpression {
  let left = bitAnd(scanner, literalFactory)

  const operators = ['^']
  const operatorMap: Record<string, BinaryExpression['operator']> = {
    '^': 'bit-xor',
  } as const

  while (true) {
    const matchedOperator = matchOperators(scanner, operators)

    if (matchedOperator !== null && matchedOperator in operatorMap) {
      const operator = operatorMap[matchedOperator]
      const right = bitAnd(scanner, literalFactory)
      left = { type: 'binary', operator, left, right }
    } else {
      return left
    }
  }
}

function bitAnd(
  scanner: Scanner,
  literalFactory: LiteralFactory,
): LogicalExpression {
  let left = bitShift(scanner, literalFactory)

  const operators = ['&']
  const operatorMap: Record<string, BinaryExpression['operator']> = {
    '&': 'bit-and',
  } as const

  while (true) {
    const matchedOperator = matchOperators(scanner, operators) ?? null

    if (matchedOperator !== null && matchedOperator in operatorMap) {
      const operator = operatorMap[matchedOperator]
      const right = bitShift(scanner, literalFactory)
      left = { type: 'binary', operator, left, right }
    } else {
      return left
    }
  }
}

function bitShift(
  scanner: Scanner,
  literalFactory: LiteralFactory,
): LogicalExpression {
  let left = additive(scanner, literalFactory)

  const operators = ['>>', '<<']
  const operatorMap: Record<string, BinaryExpression['operator']> = {
    '>>': 'bit-right-shift',
    '<<': 'bit-left-shift',
  } as const

  while (true) {
    const matchedOperator = matchOperators(scanner, operators) ?? null

    if (matchedOperator !== null && matchedOperator in operatorMap) {
      const operator = operatorMap[matchedOperator]
      const right = additive(scanner, literalFactory)
      left = { type: 'binary', operator, left, right }
    } else {
      return left
    }
  }
}

function additive(
  scanner: Scanner,
  literalFactory: LiteralFactory,
): LogicalExpression {
  let left = factor(scanner, literalFactory)

  const operators = ['+', '-']
  const operatorMap: Record<string, BinaryExpression['operator']> = {
    '+': 'addition',
    '-': 'subtraction',
  } as const

  while (true) {
    const matchedOperator = matchOperators(scanner, operators) ?? null

    if (matchedOperator !== null && matchedOperator in operatorMap) {
      const operator = operatorMap[matchedOperator]
      const right = factor(scanner, literalFactory)
      left = { type: 'binary', operator, left, right }
    } else {
      return left
    }
  }
}

function factor(
  scanner: Scanner,
  literalFactory: LiteralFactory,
): LogicalExpression {
  let left = exponentiation(scanner, literalFactory)

  const operators = ['/', '*', '%']
  const operatorMap: Record<string, BinaryExpression['operator']> = {
    '*': 'multiplication',
    '/': 'division',
    '%': 'modulus',
  } as const

  while (true) {
    const matchedOperator = matchOperators(scanner, operators) ?? null

    if (matchedOperator !== null && matchedOperator in operatorMap) {
      const operator = operatorMap[matchedOperator]
      const right = exponentiation(scanner, literalFactory)
      left = { type: 'binary', operator, left, right }
    } else {
      return left
    }
  }
}

function exponentiation(
  scanner: Scanner,
  literalFactory: LiteralFactory,
): LogicalExpression {
  const left = unary(scanner, literalFactory)

  const matchedOperator = matchOperators(scanner, ['**'])

  if (matchedOperator !== null) {
    const right = exponentiation(scanner, literalFactory)
    return { type: 'binary', operator: 'exponentiation', left, right }
  } else {
    return left
  }
}

function unary(
  scanner: Scanner,
  literalFactory: LiteralFactory,
): LogicalExpression {
  const operatorMap: Record<string, UnaryExpression['operator']> = {
    '!': 'not',
    '~': 'bit-complement',
    '-': 'negate',
    not: 'not',
  } as const

  const operators: (typeof operatorMap)['string'][] = []

  while (true) {
    const operator = matchOperators(scanner, ['~', '!', '-', 'not'])

    if (operator !== null && operator in operatorMap) {
      operators.push(operatorMap[operator])
    } else {
      break
    }
  }

  let expression = value(scanner, literalFactory)
  for (const operator of operators) {
    expression = { type: 'unary', operator, expression }
  }

  return expression
}

function value(
  scanner: Scanner,
  literalFactory: LiteralFactory,
): LogicalExpression {
  const literal = match(scanner, 'literal')
  if (literal) {
    return {
      type: 'value',
      value: {
        type: 'constant',
        value: literalFactory.create(literal.value),
      },
    }
  }

  const parameter = match(scanner, 'parameter')
  if (parameter) {
    return {
      type: 'value',
      value: { type: 'parameter', name: parameter.name },
    }
  }

  const identifier = match(scanner, 'identifier')
  if (identifier) {
    const args: LogicalExpression[] = []

    if (!match(scanner, 'group-open')) {
      return {
        type: 'value',
        value: { type: 'parameter', name: identifier.identifier },
      }
    }

    if (!match(scanner, 'group-close')) {
      while (true) {
        args.push(logicalExpression(scanner, literalFactory))

        if (!match(scanner, 'separator')) break
      }
      matchOrThrow(scanner, 'group-close')
    }

    return { type: 'function', name: identifier.identifier, arguments: args }
  }

  if (match(scanner, 'group-open')) {
    if (match(scanner, 'group-close')) {
      return { type: 'value', value: { type: 'list', items: [] } }
    }

    const expression = logicalExpression(scanner, literalFactory)

    if (!match(scanner, 'separator')) {
      matchOrThrow(scanner, 'group-close')
      return expression
    }

    const items: LogicalExpression[] = [expression]

    if (!match(scanner, 'group-close')) {
      while (true) {
        items.push(logicalExpression(scanner, literalFactory))

        if (!match(scanner, 'separator')) break
        scanner.next()
      }
    }

    return { type: 'value', value: { type: 'list', items } }
  }

  throw new UnexpectedTokenError('literal', scanner.peek)
}
