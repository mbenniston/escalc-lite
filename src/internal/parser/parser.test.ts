import { expect, test } from 'vitest'
import { parse } from './parser'
import type {
  BinaryExpression,
  FunctionExpression,
  TernaryExpression,
  UnaryExpression,
  ValueExpression,
} from './logical-expression'

test('binary operator', () => {
  expect(parse('1 + 2')).toStrictEqual({
    type: 'binary',
    operator: 'addition',
    left: {
      type: 'value',
      value: {
        type: 'constant',
        value: 1,
      },
    },
    right: {
      type: 'value',
      value: {
        type: 'constant',
        value: 2,
      },
    },
  } satisfies BinaryExpression)
})

test('unary operator', () => {
  expect(parse('-2')).toStrictEqual({
    type: 'unary',
    operator: 'negate',
    expression: {
      type: 'value',
      value: {
        type: 'constant',
        value: 2,
      },
    },
  } satisfies UnaryExpression)
})

test('ternary operator', () => {
  expect(parse('false ? 1 : 2')).toStrictEqual({
    type: 'ternary',
    left: {
      type: 'value',
      value: {
        type: 'constant',
        value: false,
      },
    },
    middle: {
      type: 'value',
      value: {
        type: 'constant',
        value: 1,
      },
    },
    right: {
      type: 'value',
      value: {
        type: 'constant',
        value: 2,
      },
    },
  } satisfies TernaryExpression)
})

test('function expression', () => {
  expect(parse('MyFunctionName(1,2,false)')).toStrictEqual({
    type: 'function',
    name: 'MyFunctionName',
    arguments: [
      {
        type: 'value',
        value: {
          type: 'constant',
          value: 1,
        },
      },
      {
        type: 'value',
        value: {
          type: 'constant',
          value: 2,
        },
      },
      {
        type: 'value',
        value: {
          type: 'constant',
          value: false,
        },
      },
    ],
  } satisfies FunctionExpression)
})

test('value expression', () => {
  expect(parse('#11/06/2000#')).toStrictEqual({
    type: 'value',
    value: {
      type: 'constant',
      value: new Date(2000, 10, 6),
    },
  } satisfies ValueExpression)
})

test.each([
  '[1] < #11/06/2000#',
  '1 < 2',
  '1 > 2',
  '1 <= 2',
  '1 >= 2',
  '1 = 2',
  '1 != 2',
  '1 <> 2',
  'true && false',
  'true and false',
  'true || false',
  'true or false',
  '1  in (1,)',
  '1 not in (1,)',
  '1 + 2',
  '1 - 2',
  '1 * 2',
  '1 / 2',
  '1 % 2',
  '1 ** 2',
  'MyFunction(1,2)',
  '1 * MyParam * {MyParam} * [MyParam] * (1,2)',
  '()',
  '(1,)',
  '(1,2)',
  '!false',
  '!!false',
  '---1',
  '-1',
  '~1',
  '~~~~1',
  '1 | 2',
  '1 & 2',
  '1 ^ 2',
  '1 << 2',
  '1 >> 2',
  'Fun()',
])('complex %s', (expr) => {
  expect(parse(expr)).toMatchSnapshot(expr)
})
