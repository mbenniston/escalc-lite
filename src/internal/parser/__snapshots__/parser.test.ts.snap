// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`complex !!false > !!false 1`] = `
{
  "expression": {
    "expression": {
      "type": "value",
      "value": {
        "type": "constant",
        "value": false,
      },
    },
    "operator": "not",
    "type": "unary",
  },
  "operator": "not",
  "type": "unary",
}
`;

exports[`complex !false > !false 1`] = `
{
  "expression": {
    "type": "value",
    "value": {
      "type": "constant",
      "value": false,
    },
  },
  "operator": "not",
  "type": "unary",
}
`;

exports[`complex () > () 1`] = `
{
  "type": "value",
  "value": {
    "items": [],
    "type": "list",
  },
}
`;

exports[`complex (1,) > (1,) 1`] = `
{
  "type": "value",
  "value": {
    "items": [
      {
        "type": "value",
        "value": {
          "type": "constant",
          "value": 1,
        },
      },
    ],
    "type": "list",
  },
}
`;

exports[`complex (1,2) > (1,2) 1`] = `
{
  "type": "value",
  "value": {
    "items": [
      {
        "type": "value",
        "value": {
          "type": "constant",
          "value": 1,
        },
      },
      {
        "type": "value",
        "value": {
          "type": "constant",
          "value": 2,
        },
      },
    ],
    "type": "list",
  },
}
`;

exports[`complex > [1] < #11/06/2000# 1`] = `
{
  "left": {
    "type": "value",
    "value": {
      "name": "1",
      "type": "parameter",
    },
  },
  "operator": "less-than",
  "right": {
    "type": "value",
    "value": {
      "type": "constant",
      "value": 2000-11-06T00:00:00.000Z,
    },
  },
  "type": "binary",
}
`;

exports[`complex > 1  in (1,) 1`] = `
{
  "left": {
    "type": "value",
    "value": {
      "type": "constant",
      "value": 1,
    },
  },
  "operator": "in",
  "right": {
    "type": "value",
    "value": {
      "items": [
        {
          "type": "value",
          "value": {
            "type": "constant",
            "value": 1,
          },
        },
      ],
      "type": "list",
    },
  },
  "type": "binary",
}
`;

exports[`complex > 1 < 2 1`] = `
{
  "left": {
    "type": "value",
    "value": {
      "type": "constant",
      "value": 1,
    },
  },
  "operator": "less-than",
  "right": {
    "type": "value",
    "value": {
      "type": "constant",
      "value": 2,
    },
  },
  "type": "binary",
}
`;

exports[`complex > 1 <= 2 1`] = `
{
  "left": {
    "type": "value",
    "value": {
      "type": "constant",
      "value": 1,
    },
  },
  "operator": "less-than-equal",
  "right": {
    "type": "value",
    "value": {
      "type": "constant",
      "value": 2,
    },
  },
  "type": "binary",
}
`;

exports[`complex > 1 > 2 1`] = `
{
  "left": {
    "type": "value",
    "value": {
      "type": "constant",
      "value": 1,
    },
  },
  "operator": "more-than",
  "right": {
    "type": "value",
    "value": {
      "type": "constant",
      "value": 2,
    },
  },
  "type": "binary",
}
`;

exports[`complex > 1 >= 2 1`] = `
{
  "left": {
    "type": "value",
    "value": {
      "type": "constant",
      "value": 1,
    },
  },
  "operator": "more-than-equal",
  "right": {
    "type": "value",
    "value": {
      "type": "constant",
      "value": 2,
    },
  },
  "type": "binary",
}
`;

exports[`complex > 1 not in (1,) 1`] = `
{
  "left": {
    "type": "value",
    "value": {
      "type": "constant",
      "value": 1,
    },
  },
  "operator": "not-in",
  "right": {
    "type": "value",
    "value": {
      "items": [
        {
          "type": "value",
          "value": {
            "type": "constant",
            "value": 1,
          },
        },
      ],
      "type": "list",
    },
  },
  "type": "binary",
}
`;

exports[`complex > true && false 1`] = `
{
  "left": {
    "type": "value",
    "value": {
      "type": "constant",
      "value": true,
    },
  },
  "operator": "and",
  "right": {
    "type": "value",
    "value": {
      "type": "constant",
      "value": false,
    },
  },
  "type": "binary",
}
`;

exports[`complex > true || false 1`] = `
{
  "left": {
    "type": "value",
    "value": {
      "type": "constant",
      "value": true,
    },
  },
  "operator": "or",
  "right": {
    "type": "value",
    "value": {
      "type": "constant",
      "value": false,
    },
  },
  "type": "binary",
}
`;

exports[`complex > true and false 1`] = `
{
  "left": {
    "type": "value",
    "value": {
      "type": "constant",
      "value": true,
    },
  },
  "operator": "and",
  "right": {
    "type": "value",
    "value": {
      "type": "constant",
      "value": false,
    },
  },
  "type": "binary",
}
`;

exports[`complex > true or false 1`] = `
{
  "left": {
    "type": "value",
    "value": {
      "type": "constant",
      "value": true,
    },
  },
  "operator": "or",
  "right": {
    "type": "value",
    "value": {
      "type": "constant",
      "value": false,
    },
  },
  "type": "binary",
}
`;

exports[`complex [1] < #11/06/2000# > [1] < #11/06/2000# 1`] = `
{
  "left": {
    "type": "value",
    "value": {
      "name": "1",
      "type": "parameter",
    },
  },
  "operator": "less-than",
  "right": {
    "type": "value",
    "value": {
      "type": "constant",
      "value": 2000-11-06T00:00:00.000Z,
    },
  },
  "type": "binary",
}
`;

exports[`complex ~~~~1 > ~~~~1 1`] = `
{
  "expression": {
    "expression": {
      "expression": {
        "expression": {
          "type": "value",
          "value": {
            "type": "constant",
            "value": 1,
          },
        },
        "operator": "bit-complement",
        "type": "unary",
      },
      "operator": "bit-complement",
      "type": "unary",
    },
    "operator": "bit-complement",
    "type": "unary",
  },
  "operator": "bit-complement",
  "type": "unary",
}
`;

exports[`complex ~1 > ~1 1`] = `
{
  "expression": {
    "type": "value",
    "value": {
      "type": "constant",
      "value": 1,
    },
  },
  "operator": "bit-complement",
  "type": "unary",
}
`;

exports[`complex ---1 > ---1 1`] = `
{
  "expression": {
    "expression": {
      "expression": {
        "type": "value",
        "value": {
          "type": "constant",
          "value": 1,
        },
      },
      "operator": "negate",
      "type": "unary",
    },
    "operator": "negate",
    "type": "unary",
  },
  "operator": "negate",
  "type": "unary",
}
`;

exports[`complex -1 > -1 1`] = `
{
  "expression": {
    "type": "value",
    "value": {
      "type": "constant",
      "value": 1,
    },
  },
  "operator": "negate",
  "type": "unary",
}
`;

exports[`complex 1  in (1,) > 1  in (1,) 1`] = `
{
  "left": {
    "type": "value",
    "value": {
      "type": "constant",
      "value": 1,
    },
  },
  "operator": "in",
  "right": {
    "type": "value",
    "value": {
      "items": [
        {
          "type": "value",
          "value": {
            "type": "constant",
            "value": 1,
          },
        },
      ],
      "type": "list",
    },
  },
  "type": "binary",
}
`;

exports[`complex 1 != 2 > 1 != 2 1`] = `
{
  "left": {
    "type": "value",
    "value": {
      "type": "constant",
      "value": 1,
    },
  },
  "operator": "not-equals",
  "right": {
    "type": "value",
    "value": {
      "type": "constant",
      "value": 2,
    },
  },
  "type": "binary",
}
`;

exports[`complex 1 % 2 > 1 % 2 1`] = `
{
  "left": {
    "type": "value",
    "value": {
      "type": "constant",
      "value": 1,
    },
  },
  "operator": "modulus",
  "right": {
    "type": "value",
    "value": {
      "type": "constant",
      "value": 2,
    },
  },
  "type": "binary",
}
`;

exports[`complex 1 & 2 > 1 & 2 1`] = `
{
  "left": {
    "type": "value",
    "value": {
      "type": "constant",
      "value": 1,
    },
  },
  "operator": "bit-and",
  "right": {
    "type": "value",
    "value": {
      "type": "constant",
      "value": 2,
    },
  },
  "type": "binary",
}
`;

exports[`complex 1 && not !true > 1 && not !true 1`] = `
{
  "left": {
    "type": "value",
    "value": {
      "type": "constant",
      "value": 1,
    },
  },
  "operator": "and",
  "right": {
    "expression": {
      "expression": {
        "type": "value",
        "value": {
          "type": "constant",
          "value": true,
        },
      },
      "operator": "not",
      "type": "unary",
    },
    "operator": "not",
    "type": "unary",
  },
  "type": "binary",
}
`;

exports[`complex 1 * 2 > 1 * 2 1`] = `
{
  "left": {
    "type": "value",
    "value": {
      "type": "constant",
      "value": 1,
    },
  },
  "operator": "multiplication",
  "right": {
    "type": "value",
    "value": {
      "type": "constant",
      "value": 2,
    },
  },
  "type": "binary",
}
`;

exports[`complex 1 * MyParam * {MyParam} * [MyParam] * (1,2) > 1 * MyParam * {MyParam} * [MyParam] * (1,2) 1`] = `
{
  "left": {
    "left": {
      "left": {
        "left": {
          "type": "value",
          "value": {
            "type": "constant",
            "value": 1,
          },
        },
        "operator": "multiplication",
        "right": {
          "type": "value",
          "value": {
            "name": "MyParam",
            "type": "parameter",
          },
        },
        "type": "binary",
      },
      "operator": "multiplication",
      "right": {
        "type": "value",
        "value": {
          "name": "MyParam",
          "type": "parameter",
        },
      },
      "type": "binary",
    },
    "operator": "multiplication",
    "right": {
      "type": "value",
      "value": {
        "name": "MyParam",
        "type": "parameter",
      },
    },
    "type": "binary",
  },
  "operator": "multiplication",
  "right": {
    "type": "value",
    "value": {
      "items": [
        {
          "type": "value",
          "value": {
            "type": "constant",
            "value": 1,
          },
        },
        {
          "type": "value",
          "value": {
            "type": "constant",
            "value": 2,
          },
        },
      ],
      "type": "list",
    },
  },
  "type": "binary",
}
`;

exports[`complex 1 ** 2 > 1 ** 2 1`] = `
{
  "left": {
    "type": "value",
    "value": {
      "type": "constant",
      "value": 1,
    },
  },
  "operator": "exponentiation",
  "right": {
    "type": "value",
    "value": {
      "type": "constant",
      "value": 2,
    },
  },
  "type": "binary",
}
`;

exports[`complex 1 + 2 > 1 + 2 1`] = `
{
  "left": {
    "type": "value",
    "value": {
      "type": "constant",
      "value": 1,
    },
  },
  "operator": "addition",
  "right": {
    "type": "value",
    "value": {
      "type": "constant",
      "value": 2,
    },
  },
  "type": "binary",
}
`;

exports[`complex 1 / 2 > 1 / 2 1`] = `
{
  "left": {
    "type": "value",
    "value": {
      "type": "constant",
      "value": 1,
    },
  },
  "operator": "division",
  "right": {
    "type": "value",
    "value": {
      "type": "constant",
      "value": 2,
    },
  },
  "type": "binary",
}
`;

exports[`complex 1 < 2 > 1 < 2 1`] = `
{
  "left": {
    "type": "value",
    "value": {
      "type": "constant",
      "value": 1,
    },
  },
  "operator": "less-than",
  "right": {
    "type": "value",
    "value": {
      "type": "constant",
      "value": 2,
    },
  },
  "type": "binary",
}
`;

exports[`complex 1 << 2 > 1 << 2 1`] = `
{
  "left": {
    "type": "value",
    "value": {
      "type": "constant",
      "value": 1,
    },
  },
  "operator": "bit-left-shift",
  "right": {
    "type": "value",
    "value": {
      "type": "constant",
      "value": 2,
    },
  },
  "type": "binary",
}
`;

exports[`complex 1 <= 2 > 1 <= 2 1`] = `
{
  "left": {
    "type": "value",
    "value": {
      "type": "constant",
      "value": 1,
    },
  },
  "operator": "less-than-equal",
  "right": {
    "type": "value",
    "value": {
      "type": "constant",
      "value": 2,
    },
  },
  "type": "binary",
}
`;

exports[`complex 1 <> 2 > 1 <> 2 1`] = `
{
  "left": {
    "type": "value",
    "value": {
      "type": "constant",
      "value": 1,
    },
  },
  "operator": "not-equals",
  "right": {
    "type": "value",
    "value": {
      "type": "constant",
      "value": 2,
    },
  },
  "type": "binary",
}
`;

exports[`complex 1 = 2 > 1 = 2 1`] = `
{
  "left": {
    "type": "value",
    "value": {
      "type": "constant",
      "value": 1,
    },
  },
  "operator": "equals",
  "right": {
    "type": "value",
    "value": {
      "type": "constant",
      "value": 2,
    },
  },
  "type": "binary",
}
`;

exports[`complex 1 > 2 > 1 > 2 1`] = `
{
  "left": {
    "type": "value",
    "value": {
      "type": "constant",
      "value": 1,
    },
  },
  "operator": "more-than",
  "right": {
    "type": "value",
    "value": {
      "type": "constant",
      "value": 2,
    },
  },
  "type": "binary",
}
`;

exports[`complex 1 >= 2 > 1 >= 2 1`] = `
{
  "left": {
    "type": "value",
    "value": {
      "type": "constant",
      "value": 1,
    },
  },
  "operator": "more-than-equal",
  "right": {
    "type": "value",
    "value": {
      "type": "constant",
      "value": 2,
    },
  },
  "type": "binary",
}
`;

exports[`complex 1 >> 2 > 1 >> 2 1`] = `
{
  "left": {
    "type": "value",
    "value": {
      "type": "constant",
      "value": 1,
    },
  },
  "operator": "bit-right-shift",
  "right": {
    "type": "value",
    "value": {
      "type": "constant",
      "value": 2,
    },
  },
  "type": "binary",
}
`;

exports[`complex 1 ^ 2 > 1 ^ 2 1`] = `
{
  "left": {
    "type": "value",
    "value": {
      "type": "constant",
      "value": 1,
    },
  },
  "operator": "bit-xor",
  "right": {
    "type": "value",
    "value": {
      "type": "constant",
      "value": 2,
    },
  },
  "type": "binary",
}
`;

exports[`complex 1 | 2 > 1 | 2 1`] = `
{
  "left": {
    "type": "value",
    "value": {
      "type": "constant",
      "value": 1,
    },
  },
  "operator": "bit-or",
  "right": {
    "type": "value",
    "value": {
      "type": "constant",
      "value": 2,
    },
  },
  "type": "binary",
}
`;

exports[`complex 1 - 2 > 1 - 2 1`] = `
{
  "left": {
    "type": "value",
    "value": {
      "type": "constant",
      "value": 1,
    },
  },
  "operator": "subtraction",
  "right": {
    "type": "value",
    "value": {
      "type": "constant",
      "value": 2,
    },
  },
  "type": "binary",
}
`;

exports[`complex 1 in !true > 1 in !true 1`] = `
{
  "left": {
    "type": "value",
    "value": {
      "type": "constant",
      "value": 1,
    },
  },
  "operator": "in",
  "right": {
    "expression": {
      "type": "value",
      "value": {
        "type": "constant",
        "value": true,
      },
    },
    "operator": "not",
    "type": "unary",
  },
  "type": "binary",
}
`;

exports[`complex 1 not in (1,) > 1 not in (1,) 1`] = `
{
  "left": {
    "type": "value",
    "value": {
      "type": "constant",
      "value": 1,
    },
  },
  "operator": "not-in",
  "right": {
    "type": "value",
    "value": {
      "items": [
        {
          "type": "value",
          "value": {
            "type": "constant",
            "value": 1,
          },
        },
      ],
      "type": "list",
    },
  },
  "type": "binary",
}
`;

exports[`complex Fun() > Fun() 1`] = `
{
  "arguments": [],
  "name": "Fun",
  "type": "function",
}
`;

exports[`complex MyFunction(1,2) > MyFunction(1,2) 1`] = `
{
  "arguments": [
    {
      "type": "value",
      "value": {
        "type": "constant",
        "value": 1,
      },
    },
    {
      "type": "value",
      "value": {
        "type": "constant",
        "value": 2,
      },
    },
  ],
  "name": "MyFunction",
  "type": "function",
}
`;

exports[`complex true && false > true && false 1`] = `
{
  "left": {
    "type": "value",
    "value": {
      "type": "constant",
      "value": true,
    },
  },
  "operator": "and",
  "right": {
    "type": "value",
    "value": {
      "type": "constant",
      "value": false,
    },
  },
  "type": "binary",
}
`;

exports[`complex true || false > true || false 1`] = `
{
  "left": {
    "type": "value",
    "value": {
      "type": "constant",
      "value": true,
    },
  },
  "operator": "or",
  "right": {
    "type": "value",
    "value": {
      "type": "constant",
      "value": false,
    },
  },
  "type": "binary",
}
`;

exports[`complex true and false > true and false 1`] = `
{
  "left": {
    "type": "value",
    "value": {
      "type": "constant",
      "value": true,
    },
  },
  "operator": "and",
  "right": {
    "type": "value",
    "value": {
      "type": "constant",
      "value": false,
    },
  },
  "type": "binary",
}
`;

exports[`complex true or false > true or false 1`] = `
{
  "left": {
    "type": "value",
    "value": {
      "type": "constant",
      "value": true,
    },
  },
  "operator": "or",
  "right": {
    "type": "value",
    "value": {
      "type": "constant",
      "value": false,
    },
  },
  "type": "binary",
}
`;
